contracts/extensions/CoinMachine.sol

```diff
@@ -144,32 +158,47 @@ contract CoinMachine is ColonyExtension {
   function buyTokens(uint256 _numTokens) public payable notDeprecated {
     updatePeriod();
 
-    uint256 numTokens = min(min(_numTokens, maxPerPeriod - activeSold), tokensToSell);
+    require(
+      whitelist == address(0x0) || Whitelist(whitelist).approved(msg.sender),
+      "coin-machine-unauthorised"
+    );
+
+    uint256 tokenBalance = ERC20(token).balanceOf(address(this));
+    uint256 numTokens = min(min(_numTokens, maxPerPeriod - activeSold), tokenBalance);
     uint256 totalCost = wmul(numTokens, activePrice);
 
     activeIntake = add(activeIntake, totalCost);
     activeSold = add(activeSold, numTokens);
-    tokensToSell = sub(tokensToSell, numTokens);
 
     assert(activeSold <= maxPerPeriod);
 
     if (purchaseToken == address(0x0)) {
       require(msg.value >= totalCost, "coin-machine-insufficient-funds");
-      if (msg.value > totalCost) { msg.sender.transfer(msg.value - totalCost); }
+      if (msg.value > totalCost) { msg.sender.transfer(msg.value - totalCost); } // Refund any balance
+      payable(address(colony)).transfer(totalCost);
     } else {
-      ERC20Extended(purchaseToken).transferFrom(msg.sender, address(this), totalCost);
+      ERC20(purchaseToken).transferFrom(msg.sender, address(colony), totalCost);
     }
 
-    colony.mintTokensFor(msg.sender, numTokens);
+    ERC20(token).transfer(msg.sender, numTokens);
 
-    emit TokensBought(msg.sender, _numTokens, totalCost);
+    emit TokensBought(msg.sender, numTokens, totalCost);
   }
 
   /// @notice Bring the token accounting current
   function updatePeriod() public {
     uint256 currentPeriod = getCurrentPeriod();
     uint256 initialActivePeriod = activePeriod;
 
+    if (initialActivePeriod == currentPeriod) {
+      return;
+    }
+
+    // If we're out of tokens, don't update
```

Comment:
I did think about that, it sort of strikes me as similar to the distinction between a closed and open set. I'm comfortable with one step of price evolution bridging arbitrarily long periods of inactivity.

Reply:
I mean, yeah, it's one step of price evolution, but it can span an arbitrarily large number of periods, and thus cause the price to decay an arbitrarily large amount, which doesn't seem desirable.

Reply:
Right, I see what you mean. If we run out of tokens and no one interacts with the contract until tokens are added, then the next `updatePeriod` will consider the entire interval for price evolution.